File: webtau-reactjs/src/report/PerformanceReport.js
/*
 * Copyright 2019 TWO SIGMA OPEN SOURCE, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import StatusEnum from './StatusEnum'

export default class PerformanceReport {
    constructor(report) {
        this.httpCalls = report.httpCalls
        this.httpCallsById = report.httpCallsById

        this.openApiHttpCallIdsPerOperation = report.openApiHttpCallIdsPerOperation
        this.notFailedHttpCalls = withoutFailedCalls(report.httpCalls)
        this.sortedNotFailedHttpCalls = sortByLatency(this.notFailedHttpCalls)

        this._calcOverallPercentile()
        this._calcPerformancePerOperation()
    }

    _calcOverallPercentile() {
        const hasNonFailedTests = this.sortedNotFailedHttpCalls.length > 0

        this.maxLatency = hasNonFailedTests ?
            this.sortedNotFailedHttpCalls[this.sortedNotFailedHttpCalls.length - 1].elapsedTime :
            0

        this.percentile = {}

        const steps = [10, 25, 50, 75, 90, 95, 99]
        steps.forEach(percentile => {
            this.percentile[percentile] = hasNonFailedTests ?
                calcPercentile(this.sortedNotFailedHttpCalls, percentile) : 0
        })
    }

    _calcPerformancePerOperation() {
        this.performancePerOperation = sortByCount(
            this.openApiHttpCallIdsPerOperation.map(callIdsAndOperation => {
                return this._calcOperationPerformance(callIdsAndOperation)
            }))
    }

    _calcOperationPerformance({method, url, httpCallIds}) {
        const httpCalls = httpCallIds.map(id => this.httpCallsById[id])
        const sortedHttpCalls = sortByLatency(httpCalls)

        return {
            method,
            url,
            count: httpCalls.length,
            fastest: sortedHttpCalls[0].elapsedTime,
            slowest: sortedHttpCalls[sortedHttpCalls.length - 1].elapsedTime,
            percentile: {
                50: calcPercentile(sortedHttpCalls, 50),
                75: calcPercentile(sortedHttpCalls, 75)
            }
        }
    }
}

function sortByLatency(httpCalls) {
    const sorted = [...httpCalls]
    sorted.sort((a, b) => a.elapsedTime - b.elapsedTime)

    return sorted
}

function sortByCount(entries) {
    const sorted = [...entries]
    sorted.sort((a, b) => b.count - a.count)

    return sorted
}

function calcPercentile(sortedHttpCalls, percentile) {
    const idx = Math.ceil((percentile / 100.0) * sortedHttpCalls.length) - 1
    return {idx, value: sortedHttpCalls[idx].elapsedTime}
}

function withoutFailedCalls(httpCalls) {
    return httpCalls.filter(c => c.status === StatusEnum.PASSED)
}

File: webtau-core/src/main/java/org/testingisdocumenting/webtau/report/perf/PerformanceReport.java
/*
 * Copyright 2021 webtau maintainers
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.testingisdocumenting.webtau.report.perf;

import org.apache.commons.math.stat.descriptive.rank.Percentile;

import java.util.ArrayList;
import java.util.List;
import java.util.LongSummaryStatistics;
import java.util.Map;
import java.util.stream.Collectors;

public class PerformanceReport {
    private String label;
    private final List<OperationPerformance> operations;
    private final List<OperationAggregatedPerformance> aggregatedOperations;

    public PerformanceReport(String label) {
        this.label = label;
        this.operations = new ArrayList<>();
        this.aggregatedOperations = new ArrayList<>();
    }

    public String getLabel() {
        return label;
    }

    public void setLabel(String label) {
        this.label = label;
    }

    public void addOperation(OperationPerformance operation) {
        this.operations.add(operation);
    }

    public List<OperationAggregatedPerformance> getAggregatedOperations() {
        return aggregatedOperations;
    }

    public void aggregate() {
        Map<String, List<OperationPerformance>> byGroupId = operations.stream()
                .collect(Collectors.groupingBy(OperationPerformance::getGroupId));

        aggregatedOperations.clear();
        aggregatedOperations.addAll(byGroupId.values().stream()
            .map(this::aggregateSingleGroup)
                .collect(Collectors.toList()));
    }

    private OperationAggregatedPerformance aggregateSingleGroup(List<OperationPerformance> operations) {
        OperationAggregatedPerformance result = new OperationAggregatedPerformance();
        result.setGroupId(operations.get(0).getGroupId());

        LongSummaryStatistics summaryStatistics = operations
                .stream()
                .collect(Collectors.summarizingLong(OperationPerformance::getElapsedMs));

        result.setAverageMs(summaryStatistics.getAverage());
        result.setMinMs(summaryStatistics.getMin());
        result.setMaxMs(summaryStatistics.getMax());
        result.setCount(summaryStatistics.getCount());

        double[] times = operations.stream()
                .map(OperationPerformance::getElapsedMs)
                .mapToDouble(Long::doubleValue)
                .sorted()
                .toArray();

        Percentile percentile = new Percentile();
        result.setP50ms(percentile.evaluate(times, 50));
        result.setP20ms(percentile.evaluate(times, 20));
        result.setP80ms(percentile.evaluate(times, 80));
        result.setP95ms(percentile.evaluate(times, 95));
        result.setP99ms(percentile.evaluate(times, 99));

        return result;
    }
}
